File: format_seclist.py
#!/usr/bin/env python

"""
Given output from seclist, print it in a suitable format for the squidlog parsing stuff
"""

import sys
import socket
import os

import functools

from collections import defaultdict

def gethostbyname(host):
    try:
        ip = socket.gethostbyname(host)
    except socket.gaierror:
        return host
    return ip

"""
An Entry is a log line

type is either LOGIN, LOGOUT or, where we have matched up the lines SESSION

"""
class Entry(object):
    months = {
           "01": "Jan",
           "02": "Feb",
           "03": "Mar",
           "04": "Apr",
           "05": "May",
           "06": "Jun",
           "07": "Jul",
           "08": "Aug",
           "09": "Sep",
           "10": "Oct",
           "11": "Nov",
           "12": "Dec"
           }

    def __init__(self, l):
        (user, host, date, time, _, inout) = l.split()
        self.user = user
        self.host = host
        self.type = inout
        if self.type == "LOGIN":
            self.starttime = self._format_time(date, time)
            self.endtime = ""
        else:
            self.endtime = self._format_time(date, time)
            self.starttime = ""

    def __str__(self):
        if self.type == "LOGIN":
            return "%s %s LOGIN" % (self.ip, self.starttime)
        if self.type == "LOGOUT":
            return "%s %s LOGOUT" % (self.ip, self.endtime)
        return "%s %s %s" % (self.ip, self.starttime, self.endtime)

    def _format_time(self, datestr, timestr):
        (day, mth, yr) = datestr.split("/")
        return "%s %s %s %s" % (day, self.months[mth], yr, timestr)

    @property
    def ip(self):
        return gethostbyname(self.host)

"""
Pull apart an input line, then add to a dict keyed by host ip (entries_by_host)
Returns a list of lists where each list is a list of log items for a single host
"""
def parse_and_match(f):
    entries = []
    entries_by_host = defaultdict(list)
    def groupFn(x):
        entries_by_host[x.ip] = entries_by_host[x.ip] + [x]
    for l in f:
        if l.rstrip() == "":
            continue
        entries.append(Entry(l))
    map (groupFn, entries)
    return entries_by_host.values()

"""
Takes a list of lines, where the list has already been filtered to so
each line is for the same host.

Matches login/logout pairs and returns a list of Events with both startime and
endtime
"""
def match_entries(lines):
    results = []
    matched_lines = []
    to_ignore = False
    ignored_lines = set()
    # filter the list of events to deal with missing data
    for index, item in enumerate(lines):
        if to_ignore:
            ignored_lines.add(item)
            to_ignore = False
            continue
        if index < len(lines) - 1:
            if item.type != lines[index+1].type:
                matched_lines.append(item)
                continue
            if item.type == "LOGIN":  # 2 LOGIN records. Ignore the first one
                ignored_lines.add(item)
                continue
            else:                     # 2 LOGOUT records. Take the first, ignore the second
                matched_lines.append(item)
                to_ignore = True
                continue
        else: # last item
            if lines[index-1].type == item.type:
                # Either last two entries don't pair up or only one item
                ignored_lines.add(lines[index-1])
                ignored_lines.add(item)
                if matched_lines:
                    matched_lines.pop()
            elif item.type == "LOGIN":
                # lonely LOGIN, ignore
                ignored_lines.add(item)
            else:
                matched_lines.append(item)
    for item in ignored_lines:
        print "Ignored line: ", item
    for (lin, lout) in zip(matched_lines[::2], matched_lines[1::2]):
        lin.endtime = lout.endtime
        lin.type = "SESSION"
        results.append(lin)
    return results

if __name__ == '__main__':
    if len(sys.argv) != 2:
        print "Usage: format_seclist.py <file>"
        sys.exit(1)
    try:
        f = open(sys.argv[1])
    except Exception, e:
        print "Can't open file %s: %s" % (sys.argv[1], e)
    for e in map(match_entries, parse_and_match(f)):
        for x in e:
            print x

File: format_seclist.hs
{-# LANGUAGE OverloadedStrings #-}

--
-- Given output from seclist, print it in a suitable format for the squidlog parsing stuff
--

import           Control.Error        (tryIO, runEitherT)
import           Data.Either          (Either(..), lefts, rights)
import           Data.List            (groupBy, partition)
import qualified Data.List as L
import qualified Data.Map as M
import           Data.Map             (Map, (!))
import           Data.Maybe           (catMaybes)
import           Network.BSD          (HostEntry(..), getHostByName)
import           Network.Socket       (inet_ntoa)
import           Safe                 (abort)
import           System.Environment   (getArgs)

months = M.fromList [
                ("01", "Jan"),
                ("02", "Feb"),
                ("03", "Mar"),
                ("04", "Apr"),
                ("05", "May"),
                ("06", "Jun"),
                ("07", "Jul"),
                ("08", "Aug"),
                ("09", "Sep"),
                ("10", "Oct"),
                ("11", "Nov"),
                ("12", "Dec")
            ]

data Line = Line {
    luser   :: String,
    lipaddr :: String,
    ldate   :: String,
    ltime   :: String,
    ltype   :: String
}

instance Show Line where
    show l = lipaddr l ++ " " ++ ldate l ++ " " ++ ltime l ++ " " ++ ltype l

data Event = Event {
    user      :: String,
    ipaddr    :: String,
    startDate :: String,
    endDate   :: String,
    startTime :: String,
    endTime   :: String
}

instance Show Event where
    show e = ipaddr e ++ " " ++ formatTime (startDate e) (startTime e)
                      ++ " " ++ formatTime (endDate e) (endTime e)
      where
        formatTime d t = format_date ++ " " ++ t
          where
            format_date = day ++ " " ++ mth ++ " " ++ yr
            day = take 2 d
            mth = months ! take 2 (drop 3 d)
            yr  = take 4 $ drop 6 d

getHost::String -> IO String
getHost name = do
        res <- runEitherT $ tryIO $ getHostByName name
        case res of
            Left  _ -> return name
            Right v -> inet_ntoa $ head (hostAddresses v)

-- very simple cmdargs handler
parseArgs::IO String
parseArgs = do
    args <- getArgs
    if length args == 1
        then return (head args)
        else abort "Usage: format_seclist [file1]"

-- Convert list of strings to parsed items
parseInput::[String] -> IO (Maybe Line)
parseInput [user, host, date, time, _, inout] = do
    ipaddr <- getHost host
    return $ Just (Line user ipaddr date time inout)
parseInput [] = return Nothing

-- Given a list of lines all from the same host
-- match up login/logout times return a pair
-- ([matched events], [unmatched lines)]
matchEntries::[Line] -> ([Event], [Line])
matchEntries ls = (rights matches, lefts matches)
  where
    matches = catMaybes $ zipWith matchPairs ls (tail ls) ++ [matchLastEntry ls]
    matchPairs e1 e2 = case (ltype e1, ltype e2) of
                            ("LOGIN", "LOGOUT")  -> Just (Right $ makeEvent e1 e2)
                            ("LOGIN", "LOGIN")   -> Just (Left e1)
                            ("LOGOUT", "LOGIN")  -> Nothing
                            ("LOGOUT", "LOGOUT") -> Just (Left e2)
    -- check if we ignored the last entry or not
    matchLastEntry []  = Nothing
    matchLastEntry [x] = Just (Left x)
    matchLastEntry ls  = case (ltype e1, ltype e2) of
                                 ("LOGIN", "LOGIN")   -> Just (Left e2)
                                 ("LOGOUT", "LOGIN")  -> Just (Left e2)
                                  -- other cases already covered in matchPairs
                                 (_      , _)         -> Nothing
      where
         e1 = last $ init ls
         e2 = last ls

    makeEvent e1 e2 = Event (luser e1) (lipaddr e1) (ldate e1)
                                                (ldate e2) (ltime e1) (ltime e2)

-- group the list of entries by host/ip
groupByHost::[Line] -> [[Line]]
groupByHost parsedLines = M.elems $ foldl f M.empty parsedLines
  where
    f m x  = M.insertWith (flip (L.++) ) (lipaddr x) [x] m

main = do
    s <- parseArgs >>= readFile
    let parts = map words (lines s)
    parsedLines <- mapM parseInput parts
    let linesByHost = groupByHost $ catMaybes parsedLines
    mapM_ (printResults . matchEntries) linesByHost
  where
    printResults (events, ignored) = do
        mapM_ (\e -> putStrLn $ "Ignored line: " ++ show e) ignored
        mapM_ print events
